scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala

```diff
@@ -1982,32 +2051,135 @@ class FormatOps(
         }
         if (ok) Some(new OptionalBraces(tree, true)) else None
       }
-      ftMeta.leftOwner match {
-        case t: Term.Match => result(t, t.cases)
-        case t: Type.Match => result(t, t.cases)
-        case _ => None
+      getMissingBrace(ftMeta).flatMap { _ =>
+        ftMeta.leftOwner match {
+          case t: Term.Match => result(t, t.cases)
+          case t: Type.Match => result(t, t.cases)
+          case _ => None
+        }
       }
     }
   }
 
+  // Optional braces after any token that can start indentation:
+  // then
+  object OptionalBracesDo {
+    def unapply(
+        ftMeta: FormatToken.Meta
+    )(implicit style: ScalafmtConfig): Option[OptionalBraces] =
+      getMissingBrace(ftMeta).flatMap { nft =>
+        (ftMeta.leftOwner match {
+          case t: Term.Do => Some(t.body -> true)
+          case t: Term.While => Some(t.body -> false)
+          case t: Term.For => Some(t.body -> false)
+          case _ => None
+        }).map { case (body, allowMain) =>
+          val singleStat = !isTreeMultiStatBlock(body)
+          val forceNL = !singleStat || shouldBreakInOptionalBraces(nft)
+          new OptionalBraces(body, forceNL, allowMain && singleStat)
+        }
+      }
+  }
+
+  // Optional braces after any token that can start indentation:
+  // then
+  object OptionalBracesThen {
+    def unapply(
+        ftMeta: FormatToken.Meta
+    )(implicit style: ScalafmtConfig): Option[OptionalBraces] =
+      getMissingBrace(ftMeta).flatMap { nft =>
+        ftMeta.leftOwner match {
+          case t: Term.If =>
+            def nestedIf(x: Term.If) = {
+              val forceNL = shouldBreakInOptionalBraces(nft) ||
+                !ifWithoutElse(t) && existsIfWithoutElse(x)
+              Some(new OptionalBraces(x, forceNL))
+            }
+            t.thenp match {
+              case x: Term.If => nestedIf(x)
+              case Term.Block(List(x: Term.If)) => nestedIf(x)
+              case x =>
+                val forceNL =
+                  isTreeMultiStatBlock(x) || shouldBreakInOptionalBraces(nft)
+                Some(new OptionalBraces(x, forceNL))
+            }
+          case _ => None
+        }
+      }
+  }
+
+  // Optional braces after any token that can start indentation:
+  // else
+  object OptionalBracesElse {
+    def unapply(
+        ftMeta: FormatToken.Meta
+    )(implicit style: ScalafmtConfig): Option[OptionalBraces] =
+      getMissingBrace(ftMeta).flatMap { nft =>
```

Comment:
It seems that we need to check `getMissingBrace(ftMeta)` on each of the new unapplies, which makes us do it over and over again. I wonder whether we should do it all in a single `OptionalBraces` object? Looks like we should be able to merge that. Or am I too worried about performance impact that does not exist there?

Reply:
good comment. i think it's time to go back to a single matcher. i will send an update in the morning.